src/algebraic_geometry/EllipticCurve/point.lean

```diff
@@ -0,0 +1,327 @@
+@[simp] def add_y' : K := add_L h₁ h₂ * (add_x h₁ h₂ - x₁) + y₁
+
+/-- The $y$-coordinate of the addition of two affine points after applying the final negation. -/
+@[simp] def add_y : K := -add_y' h₁ h₂ - ↑E.a₁ * add_x h₁ h₂ - ↑E.a₃
+
+/-- The negation of the addition of two affine points lies on `E`. -/
+lemma weierstrass_add' (hx : x₁ ≠ x₂) : E.weierstrass (add_x h₁ h₂) (add_y' h₁ h₂) :=
+begin
+  rw [weierstrass, ← sub_eq_zero],
+  convert_to
+    add_L h₁ h₂ * (add_L h₁ h₂ * (add_L h₁ h₂ * (add_L h₁ h₂ * (x₁ - x₂) * -1
+                                                  + (-↑E.a₁ * x₁ + 2 * ↑E.a₁ * x₂ + 2 * y₁ + ↑E.a₃))
+                                  + (x₁ ^ 2 - 2 * x₁ * x₂ - 2 * x₂ ^ 2 + ↑E.a₁ ^ 2 * x₂
+                                      - 2 * ↑E.a₂ * x₂ + 3 * ↑E.a₁ * y₁ + ↑E.a₁ * ↑E.a₃ - ↑E.a₄))
+                    + (-↑E.a₁ * x₁ ^ 2 - 3 * ↑E.a₁ * x₁ * x₂ - 4 * x₁ * y₁ - 2 * ↑E.a₁ * x₂ ^ 2
+                        - 2 * x₂ * y₁ - ↑E.a₁ * ↑E.a₂ * x₁ - 2 * ↑E.a₃ * x₁ - 2 * ↑E.a₁ * ↑E.a₂ * x₂
+                        - ↑E.a₃ * x₂ + ↑E.a₁ ^ 2 * y₁ - 2 * ↑E.a₂ * y₁ - ↑E.a₁ * ↑E.a₄
+                        - ↑E.a₂ * ↑E.a₃))
+    + (x₁ ^ 3 + 3 * x₁ ^ 2 * x₂ + 3 * x₁ * x₂ ^ 2 + x₂ ^ 3 + 2 * ↑E.a₂ * x₁ ^ 2
+        + 4 * ↑E.a₂ * x₁ * x₂ - ↑E.a₁ * x₁ * y₁ + 2 * ↑E.a₂ * x₂ ^ 2 - ↑E.a₁ * x₂ * y₁ + y₁ ^ 2
+        + ↑E.a₂ ^ 2 * x₁ + ↑E.a₄ * x₁ + ↑E.a₂ ^ 2 * x₂ + ↑E.a₄ * x₂ - ↑E.a₁ * ↑E.a₂ * y₁
+        + ↑E.a₃ * y₁ + ↑E.a₂ * ↑E.a₄ - ↑E.a₆) = 0,
+  { simp only [add_x, add_y'], ring1 },
+  field_simp [sub_ne_zero_of_ne hx],
+  rw [weierstrass] at h₁ h₂,
+  linear_combination
+    -((x₁ - x₂) ^ 2 * (x₁ + 2 * x₂ + ↑E.a₂) - (x₁ - x₂) * (y₁ - y₂) * ↑E.a₁ - (y₁ - y₂) ^ 2) * h₁
+    + ((x₁ - x₂) ^ 2 * (2 * x₁ + x₂ + ↑E.a₂) - (x₁ - x₂) * (y₁ - y₂) * ↑E.a₁ - (y₁ - y₂) ^ 2) * h₂
+    with { normalization_tactic := `[ring1] }
+end
+
+/-- The addition of two affine points lies on `E`. -/
+lemma weierstrass_add (hx : x₁ ≠ x₂) : E.weierstrass (add_x h₁ h₂) (add_y h₁ h₂) :=
+weierstrass_neg $ weierstrass_add' h₁ h₂ hx
+
+private lemma y_eq_of_y_ne (hx : x₁ = x₂) (hy : y₁ ≠ neg_y h₂) : y₁ = y₂ :=
+begin
+  rw [weierstrass] at h₂,
+  rw [weierstrass, hx, ← h₂, ← sub_eq_zero] at h₁,
+  apply eq_of_sub_eq_zero ∘ eq_zero_of_ne_zero_of_mul_right_eq_zero (sub_ne_zero.mpr hy),
+  convert h₁,
+  rw [neg_y],
+  ring1
+end
+
+private lemma y_ne_of_y_ne (hx : x₁ = x₂) (hy : y₁ ≠ neg_y h₂) : y₁ ≠ neg_y h₁ :=
+by { convert hy, exact y_eq_of_y_ne h₁ h₂ hx hy }
+
+omit h₁ h₂
+
+open_locale classical
+
+/-- Addition in `E⟮K⟯`. -/
+@[simp] noncomputable def add : E⟮K⟯ → E⟮K⟯ → E⟮K⟯
+| 0               P               := P
+| P               0               := P
+| (some x₁ y₁ h₁) (some x₂ y₂ h₂) :=
+if hx : x₁ = x₂ then if hy : y₁ = neg_y h₂ then 0
+else some _ _ $ weierstrass_dbl h₁ $ y_ne_of_y_ne h₁ h₂ hx hy
+else some _ _ $ weierstrass_add h₁ h₂ hx
+
+noncomputable instance : has_add E⟮K⟯ := ⟨add⟩
+
+@[simp] lemma add_def (P Q : E⟮K⟯) : P + Q = add P Q := rfl
```

Comment:
I don't think you want this as a (global) simp lemma. The simp lemmas below will not fire as the LHS will be rewritten first.